File: nnerik/day02.py
# Read input
input = open(0).read().strip().split("\n")
reports = [[int(level) for level in report.split()] for report in input]


# Part 1
def is_safe(report):
    sign = -1 if report[0] > report[1] else 1
    for i in range(len(report) - 1):
        if not 1 <= sign * (report[i + 1] - report[i]) <= 3:
            return False
    return True


print("Part 1:", sum(1 for report in reports if is_safe(report)))


# Part 2
def is_safe_2(report):
    for skip in range(len(report)):
        if is_safe(report[:skip] + report[skip + 1 :]):
            return True
    return False


print("Part 2:", sum(1 for report in reports if is_safe_2(report)))

File: fredriv/day02/day02.scala
import scala.io.Source
import java.lang.Math

@main def day02 =
    val input = Source.fromFile("input.txt").getLines.toVector
    val reports = input.map(_.split(" ").map(_.toInt))

    println("Part 1: " + reports.count(isSafe))
    println("Part 2: " + reports.count(isSafeWithDampener))

def isSafe(report: Array[Int]): Boolean =
    val pairs = report.sliding(2).map(p => (p(0), p(1))).toList
    pairs.forall((a, b) => Math.abs(a - b) <= 3) &&
        (pairs.forall((a, b) => a > b) || pairs.forall((a, b) => a < b))

def isSafeWithDampener(report: Array[Int]): Boolean =
    if isSafe(report) then true
    else (0 until report.length).exists { i =>
        val dampened = report.take(i) ++ report.drop(i + 1)
        isSafe(dampened)
    }
